woocommerce/woocommerce-ios
# Refactor `OrderSearchViewController` into a generic search UI

Fixes #1302 

Update release notes:


## Changes
- Note: thanks a bunch for @ctarda's example playground and post, see if there's any part that can be improved
- Renamed the original `OrderSearchViewController` and xib to `SearchViewController`
- Updated search table view style from `grouped` to `plain` per design review in https://github.com/woocommerce/woocommerce-ios/pull/1290#pullrequestreview-290666469
- Created `SearchCommand` as an interface that provides/syncrhonizes data for rendering search UI based on a generic model
  - Added `OrderSearchCommand` that implements `SearchCommand` for Order search
- Created `SearchResultCell` as an interface that renders a generic search result model
  - Updated `OrderTableViewCell` to implement `SearchResultCell`
  - Created `OrderSearchCellViewModel` as the view model for `OrderTableViewCell` as Order search result cell since it takes not only `Order` but `OrderStatus` model to render the cell
- Moved Order dependent search logic from the original `OrderSearchViewController` to `OrderSearchCommand`
- In `OrdersViewController`, updated how Order search view controller is initialized to `SearchViewController`

## Testing
Prerequisite: the store has some orders
The goal is to maintain the same Order search behavior with the only UI change of table view style from `grouped` to `plain`
- Launch the app
- Go to "Orders" tab
- Tap the search icon 🔍 on the navbar --> the existing Orders should be shown (the same behavior as Order search before)
- Type in a search query that renders some results --> the corresponding Orders that match the query should be shown (the same behavior as Order search before)
- Type in a search query that doesn't render any results --> "No Orders found" placeholder should be shown (the same behavior as Order search before)

Note: since the search table view style is updated from `grouped` to `plain`, you might see the spinner without top margin. The spinner will be polished in https://github.com/woocommerce/woocommerce-ios/issues/1281
